src/LibSignedSettlement.sol

```diff
@@ -0,0 +1,376 @@
+    {
+        {
+            bytes calldata extraBytes;
+            (extraBytes, lastByte) = readExtraBytes(tokens, clearingPrices, trades, interactions);
+
+            if (lastByte != msg.data.length - 193) {
+                revert LibSignedSettlement__InvalidExtraParamsPartiallySigned();
+            }
+
+            assembly ("memory-safe") {
+                deadline := calldataload(lastByte)
+                r := calldataload(add(lastByte, 0x20))
+                s := calldataload(add(lastByte, 0x40))
+                v := and(calldataload(add(lastByte, 0x41)), 0xff)
+                offsets := add(lastByte, 0x61)
+            }
+
+            // validate that the subset size is <= interactions length
+            if (
+                offsets[0] > interactions[0].length || offsets[1] > interactions[1].length
+                    || offsets[2] > interactions[2].length
+            ) {
+                revert LibSignedSettlement__InvalidExtraParamsPartiallySigned();
+            }
+        }
+    }
+
+    /// @dev Reads params appended at the end, computes the digest for the fully signed message
+    ///      and returns the memory slice where the calldata is encoded for the `GPv2Settlement.settle`
+    ///      call.
+    ///
+    ///      The digest is assumed to be keccak256 hash of the following encoded data:
+    ///      `abi.encodePacked(abi.encode(tokens, clearingPrices, trades, interactions), abi.encode(deadline, solver))`
+    function getParamsDigestAndCalldataFullySigned(
+        address[] calldata tokens,
+        uint256[] calldata clearingPrices,
+        GPv2Trade.Data[] calldata trades,
+        GPv2Interaction.Data[][3] calldata interactions
+    )
+        internal
+        view
+        returns (
+            uint256 deadline,
+            uint256 r,
+            uint256 s,
+            uint256 v,
+            bytes32 digest,
+            uint256 calldataStart,
+            uint256 calldataSize
+        )
+    {
+        {
+            uint256 lastByte;
+            (deadline, r, s, v, lastByte) = readExtraParamsFullySigned(tokens, clearingPrices, trades, interactions);
+
+            assembly ("memory-safe") {
+                let freePtr := mload(0x40)
+                let dataStart := add(freePtr, 0x20)
+                // lastByte - 4(method id) + 32(deadline)
+                let nBytesToCopy := add(lastByte, 0x1c)
+                // copy all the params including the extra param `deadline` appended to the calldata
+                calldatacopy(dataStart, 0x04, nBytesToCopy)
+                // store the solver after deadline
+                mstore(add(dataStart, nBytesToCopy), caller())
+                // hash the message abi.encode(tokens, clearingPrices, trades, interactions) | deadline | solver
+                digest := keccak256(dataStart, add(nBytesToCopy, 0x20))
+                // update the freePtr
+                mstore(0x40, add(freePtr, add(nBytesToCopy, 0x20)))
```

Comment:
it is `0x20` because we allocated extra word at the start for the function selector which cannot be discarded, while the solver and deadline are already discarded.